### ISSUE TITLE
About [sam] layer.

### ISSUE THREAD

**Issue by ChenCong7375** [AlexeyAB/darknet#3708]
I noticed that you added [sam] layer in darknet. How can we use it?

cfg file with [sam]: [yolov3-tiny-sam.cfg.txt](https://github.com/AlexeyAB/darknet/files/4039980/yolov3-tiny-sam.cfg.txt)

COCO test-dev

| Model | Size | BFLOPS | Inference time, ms | AP@.5:.95 | AP@.5 | AP@.75 |
| :-- | :-: | :-: | :-: | :-: | :-: | :-: |
|  [yolo_v3_tiny_pan3 aa_ae_mixup_scale_giou (no sgdr).txt](https://github.com/AlexeyAB/darknet/files/3619445/yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou.no.sgdr.txt) | 416x416 | 8.4 | 6.4  | 18.8% | 36.8% | 17.5% |
|  [yolov3-tiny-prn.cfg.txt](https://github.com/AlexeyAB/darknet/files/3621223/yolov3-tiny-prn.cfg.txt) | 416x416 | 3.5 | 3.8 | - | 33.1% | - |
| [enet-coco.cfg.txt](https://github.com/AlexeyAB/darknet/files/3621226/enet-coco.cfg.txt) | 416x416 | 3.7 | 22.7 | - | 45.5% | - |

**Comment by LukeAI** [comment]
think it's for thundernet https://github.com/AlexeyAB/darknet/issues/3702

**Comment by WongKinYiu** [comment]
![image](https://user-images.githubusercontent.com/12152972/62534082-d1fc3b00-b87a-11e9-8665-adc6f719d3d8.png)

**Comment by ChenCong7375** [comment]
@WongKinYiu  could you please share the cfg file?

**Comment by WongKinYiu** [comment]
[yolov3-tiny-sam.cfg.txt](https://github.com/AlexeyAB/darknet/files/3479469/yolov3-tiny-sam.cfg.txt)
here u r.

**Comment by LukeAI** [comment]
@WongKinYiu Thanks for sharing another novel architecture - could you tell us a little about it? Like rough cocoAP / inf. time on an RTX?

**Comment by WongKinYiu** [comment]
https://github.com/AlexeyAB/darknet/issues/3380#issuecomment-503780307
https://github.com/AlexeyAB/darknet/issues/3380#issuecomment-517274542

**Comment by LukeAI** [comment]
Oh, I see this is the CEM + SAM + Yolov3 with 42.0% mAP@0.5 with 2.90 BFLOPs.? Sounds great, I'll see how it goes and report back. Have you done any other experimental architectures that you would be happy sharing? Do you think it might be improved by trying to use a pan-like head?

**Comment by AlexeyAB** [comment]
@LukeAI If you have a time, try to train this model (CEM + SAM + Yolov3 with 42.0% mAP@0.5 with 2.90 BFLOPs) on this dataset: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

For adding result (chart Loss & mAP, BFlops) to this table.

**Comment by ChenCong7375** [comment]
@AlexeyAB Is there any cfg file of CEM + SAM + Yolov3 ?
I will have a try.

**Comment by WongKinYiu** [comment]
[enetb0-cemsam.cfg.txt](https://github.com/AlexeyAB/darknet/files/3481663/enetb0-cemsam.cfg.txt)

Because there is no parameter can let up-sampling layer up-sample the feature maps to the size before global average pooling layer, I use max-pooling layer instead global average pooling layer in CEM.

**Comment by LukeAI** [comment]
> [yolov3-tiny-sam.cfg.txt](https://github.com/AlexeyAB/darknet/files/3479469/yolov3-tiny-sam.cfg.txt)
> here u r.

I try to train with:
./darknet detector train my_stuff/bdd100k.data my_stuff/yolov3-tiny-sam.cfg my_stuff/yolov3-tiny.conv.15 -dont_show -mjpeg_port 8090 -map -i 1`

But it immediately aborts with:

```
...
[yolo] params: iou loss: mse, iou_norm: 0.75, cls_norm: 1.00, scale_x_y: 1.00
Total BFLOPS 4.887 
 Allocate additional workspace_size = 1245.71 MB 
Loading weights from my_stuff/yolov3-tiny.conv.15...
 seen 64 
Done!
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
Resizing
608 x 608 
Resizing type 15 
Cannot resize this type of layer: File exists
darknet: ./src/utils.c:293: error: Assertion `0' failed.
....
```

Am I using the wrong pretrained weights? or something else?

**Comment by LukeAI** [comment]
training now, looking good so far.
what am I missing with random=0?
How could I add scales_x_y to this model?

**Comment by WongKinYiu** [comment]
@LukeAI 
For using scale_x_y, plz see https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

**Comment by LukeAI** [comment]
@WongKinYiu  I mean, I know that the scale models have "scale_x_y = 1.05" or something like that in the Yolo layers, I just don't really understand what an appropriate value would be. I could try with 1.05 and just see how that works? or 1.1?

**Comment by WongKinYiu** [comment]
@LukeAI to set an appropriate value, plz see https://github.com/AlexeyAB/darknet/issues/3293#issuecomment-497895809

**Comment by LukeAI** [comment]
Hi all,
Some experiments I ran a while back using the berkley deep drive dataset (slightly reduced number of classes)
Baseline:
[CEM1.cfg.txt](https://github.com/AlexeyAB/darknet/files/3559520/CEM1.cfg.txt)
![CEM1](https://user-images.githubusercontent.com/43993778/64014411-f1a11f00-cb19-11e9-880d-4b512b02769d.png)

With anchors generated from the dataset:
[CEM_with_anchors.cfg.txt](https://github.com/AlexeyAB/darknet/files/3559521/CEM_with_anchors.cfg.txt)
![CEM_with_anchors](https://user-images.githubusercontent.com/43993778/64014412-f1a11f00-cb19-11e9-9cf1-ccde1613a788.png)

Using scale_x_y=1.05
[CEM_with_scale.cfg.txt](https://github.com/AlexeyAB/darknet/files/3559522/CEM_with_scale.cfg.txt)
![CEM_with_scale](https://user-images.githubusercontent.com/43993778/64014415-f239b580-cb19-11e9-9df9-ff86ae7d1b13.png)

Using swish activations:
[CEM_with_swish.cfg.txt](https://github.com/AlexeyAB/darknet/files/3559523/CEM_with_swish.cfg.txt)
![CEM_with_swish](https://user-images.githubusercontent.com/43993778/64014416-f239b580-cb19-11e9-8ae8-9443a5af58dc.png)

For comparison, the same dataset trained with tiny_3l:
![tiny_3l](https://user-images.githubusercontent.com/43993778/64014417-f239b580-cb19-11e9-9069-802f69a30812.png)

and with tiny_pan2:
![tiny_pan2_swish_3](https://user-images.githubusercontent.com/43993778/64014418-f2d24c00-cb19-11e9-938a-82293b102015.png)

**Comment by AlexeyAB** [comment]
@LukeAI 
So `CEM, scale and swish` doesn't give significant improvements?

`tiny_pan2` is the most accuracy network?

**Comment by LukeAI** [comment]
yeah, tiny_pan2 is a good one, here's hoping for a full sized pan2 network. I didn't measure the inf. time, I guess the point of the CEM network is that is very fast whilst still being reasonably accurate?

**Comment by AlexeyAB** [comment]
@LukeAI Just add comparison table, with final accuracy, FLOPS, and inference time

**Comment by WongKinYiu** [comment]
I think the mainly improvement is from more anchors/yolo-layers.
In my experiments, yolo-v3-tiny-3l gets 5.7% higher mAP@0.5 than yolo-v3-tiny(2l) on pedestrian detection task.

**Comment by WongKinYiu** [comment]
here list some results of my backbone (evaluate on coco test-dev set):
1. with 2l (6 anchors): 45.0% mAP@0.5, 4.04 BFLOPs.
2. with 3l (9 anchors): 46.3% mAP@0.5, 5.03 BFLOPs.
3. with cem (6 anchors): 45.2% mAP@0.5, 4.81 BFLOPs.
4. with cem sam (6 anchors): 46.1% mAP@0.5, 4.90 BFLOPs.
5. with modified cem sam (9 anchors): 48.0% mAP@0.5, 4.95 BFLOPs.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

> 6. model B with **modified** cem sam (9 anchors): 48.0% mAP@0.5, 4.95 BFLOPs.

Thanks! 
What modifications did you do in 6-model?

**Comment by WongKinYiu** [comment]
@AlexeyAB Heloo, i m on a business trip, i ll share the modified cem sam tonight.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

1. i using spp instead global average pooling, it is becuz currently this repo can not support multi-scale training when using global average pooling as an intermediate layer.
2. since yolo is an one-stage object detector, i add sam layer for each feature pyramid. 

[modified-cem-sam-head.txt](https://github.com/AlexeyAB/darknet/files/3570127/modified-cem-sam-head.txt)

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks!
Did you compare Inference time (sec) for `2. model A with 3l (9 anchors): 46.3% mAP@0.5, 5.03 BFLOPs.` and `6. model B with modified cem sam (9 anchors): 48.0% mAP@0.5, 4.95 BFLOPs.` ?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello, 
sam layer is similar to scale channels layer, although it increase only <1% computation, it increase 20%~30% inference time on gpu. on cpu, they take similar inference time.

**Comment by AlexeyAB** [comment]
@WongKinYiu When you will find the best cfg-file, please share it, I will add it to this repository.

**Comment by WongKinYiu** [comment]
@AlexeyAB for best inference speed, i may share this model after discuss with my team.
![image](https://user-images.githubusercontent.com/12152972/64189325-94ed8d80-cea6-11e9-9676-7283c2e7edf2.png)
it reduces 45% number of parameter, 38% of computation, 37% CPU computation time, 19% GPU computation time, and 25% TX2 computation time, while maintaining same mAP@0.5 as yolo-v3-tiny.
this model achieves 485 fps on gtx 1080 ti (batch size = 1).

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks!

* So as I see you didn't use `CEM, SAM or Squeeze-and-Excitation blocks`

* Also you didn't use `SPP, PAN2, Assisted Excitation, Anti Aliasing, scales_x_y, GIoU` https://github.com/AlexeyAB/darknet/projects/1

Whad do you think about these features?

**Comment by WongKinYiu** [comment]
SPP and scales_x_y are very useful.
GIoU improves mAP@0.5:0.95, but drops mAP@0.5. For some cases, mAP@0.5 is more important.
PAN2 reduces 13% computation than PAN and reduces 0.2% mAP@0.5 in my experiment.
Mixup can not benefit lightweight model in my experiments.

I think Anti Aliasing will be useful too.
And I will take a look Assisted Excitation.

However, I find a strange phenomenon.
When I use repos committed before early March to train a detector, it always get better performance than using repos committed after May.
If I train the model using old repo, then valid the model using new repos. They get worse results, too.
So I do not use new features in my model.
I will try to add features that I need to old repo and test them in following one month.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

> If I train the model using old repo, then valid the model using new repos. They get worse results, too.

Maybe only the new accuracy checking function is different, and the training is just as good?
I fixed a little in mAP function.

> GIoU improves mAP@0.5:0.95, but drops mAP@0.5. For some cases, mAP@0.5 is more important.
> PAN2 reduces 13% computation than PAN and reduces 0.5% mAP@0.5 in my experiment.
> Mixup can not benefit lightweight model in my experiments.

Did you test it on MS COCO dataset?

I will add PAN3 block and new tiny model today there: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

**Comment by WongKinYiu** [comment]
@AlexeyAB 
I upload predicted bounding boxes to codalab.
And I train a same model for several times, old repo always get better results.

Yes, all of my experiment results are test on MS COCO test-dev set.

**Comment by AlexeyAB** [comment]
@WongKinYiu 
I added another one mode: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-528532293

cfg: https://github.com/AlexeyAB/darknet/files/3580764/yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou.cfg.txt

It seems it is the best cfg-file for this small dataset: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

You can try to train it on MS COCO and check the mAP if you have a time.

**Comment by AlexeyAB** [comment]
@WongKinYiu Also can you attach entire the best of yours SAM_CEM model (not only head)?
I will attach it here and close the Issue: https://github.com/AlexeyAB/darknet/issues/3702

> @AlexeyAB modified-cem-sam-head.txt
> 
> i using spp instead global average pooling, it is becuz currently this repo can not support multi-scale training when using global average pooling as an intermediate layer.
> since yolo is one-stage object detector, i add sam layer for each feature pyramid.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

Thank you for sharing a good model (yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou).

After discuss with my team, I can not share the backbone of https://github.com/AlexeyAB/darknet/issues/3708#issuecomment-526991664 currently.
I will add the modified-cem-sam-head to yolo-v3-tiny and share the cfg latter.

**Comment by WongKinYiu** [comment]
@AlexeyAB 
now training yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou (no sgdr) on coco dataset.
i ll report the result after finish training.

**Comment by AlexeyAB** [comment]
@WongKinYiu Try to increase assisted_excitation=4000 to assisted_excitation=20000 or 50000

**Comment by WongKinYiu** [comment]
| Model | Size | AP@.5:.95 | AP@.5 | AP@.75 |
| :-- | :-: | :-: | :-: | :-: |
|  [yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou (no sgdr).txt](https://github.com/AlexeyAB/darknet/files/3619445/yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou.no.sgdr.txt) | 416x416 | 18.8% | 36.8% | 17.5% |

**Comment by AlexeyAB** [comment]
COCO test-dev

| Model | Size | BFLOPS | Inference time, ms | AP@.5:.95 | AP@.5 | AP@.75 |
| :-- | :-: | :-: | :-: | :-: | :-: | :-: |
|  [yolo_v3_tiny_pan3 aa_ae_mixup_scale_giou (no sgdr).txt](https://github.com/AlexeyAB/darknet/files/3619445/yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou.no.sgdr.txt) | 416x416 | 8.4 | 6.4  | 18.8% | 36.8% | 17.5% |
|  [yolov3-tiny-prn.cfg.txt](https://github.com/AlexeyAB/darknet/files/3621223/yolov3-tiny-prn.cfg.txt) | 416x416 | 3.5 | 3.8 | - | 33.1% | - |
| [enet-coco.cfg.txt](https://github.com/AlexeyAB/darknet/files/3621226/enet-coco.cfg.txt) | 416x416 | 3.7 | 22.7 | - | 45.5% | - |

**Comment by nyj-ocean** [comment]
@AlexeyAB 
I download the `latest repo` and set like following in `makefile`

> GPU=1
CUDNN=1
CUDNN_HALF=1
OPENCV=1
AVX=0
OPENMP=0
LIBSO=0
ZED_CAMERA=0

Then use `yolov3-tiny-sam.cfg.txt` in https://github.com/AlexeyAB/darknet/issues/3708#issuecomment-519329865 to train with my own dataset
But meet error like following

> Total BFLOPS 4.883 
 Allocate additional workspace_size = 52.43 MB 
Loading weights from /home/gc/4-images/9.18/darknet/yolov3-tiny.conv.15...
 seen 64 
Done! Loaded 23 layers from weights-file 
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
 If error occurs - run training with flag: -dont_show 
Resizing
608 x 608 
Resizing type 16 
Cannot resize this type of layer: 
darknet: ./src/utils.c:297：error:

**Comment by WongKinYiu** [comment]
@nyj-ocean 

yes, you should modify the resize function of `sam` layer, or you can only train it with random=1.

**Comment by nyj-ocean** [comment]
@WongKinYiu 
`sam layers` could not train with `Multi-Scale(random=1)`,is it right?

**Comment by WongKinYiu** [comment]
@nyj-ocean 

yes.

just add the case of resize function of `sam` layer in [network.c](https://github.com/AlexeyAB/darknet/blob/master/src/network.c).
it already defined in [sam_layer.c](https://github.com/AlexeyAB/darknet/blob/master/src/sam_layer.c).
so you can simply include it and just need a little bit modification.

**Comment by nyj-ocean** [comment]
@WongKinYiu 
Thanks a lot

**Comment by nyj-ocean** [comment]
@AlexeyAB 
I notice another module:`CBAM: Convolutional Block Attention Module`
![图片](https://user-images.githubusercontent.com/55009815/72063505-dab6d880-3314-11ea-8d41-4df6eaefaabb.png)
![图片](https://user-images.githubusercontent.com/55009815/72063623-1356b200-3315-11ea-8a06-c3cea907069a.png)
![图片](https://user-images.githubusercontent.com/55009815/72063647-1f427400-3315-11ea-9a8d-742984ecce35.png)
![图片](https://user-images.githubusercontent.com/55009815/72063721-4a2cc800-3315-11ea-881a-3e80bfc060dd.png)

Is there a need to add `CBAM Module` to this repo?

[CBAM: Convolutional Block Attention Module.pdf](https://github.com/AlexeyAB/darknet/files/4040116/CBAM.Convolutional.Block.Attention.Module.pdf)
code:https://github.com/Jongchan/attention-module

**Comment by WongKinYiu** [comment]
The kernel function of CAM module and SAM module are SE(squeeze-and-excitation) and SAM, it already supported by this repo.

**Comment by AlexeyAB** [comment]
I added resizing (`random=1`) for [sam] layers.

**Comment by 924175302** [comment]
@nyj-ocean 
Hello, have you tried the CBAM module in YOLOV4?

**Comment by 924175302** [comment]
@WongKinYiu
You mentioned that this repo already supports the SE module, I don’t find the relevant code and how to use it, can you help me answer it, thank you

**Comment by WongKinYiu** [comment]
Squeeze-and-Excitation blocks (layers: `[avgpool]`->`[conv]`->`[conv]`->`[scale_channels]`)

**Comment by AlexeyAB** [comment]
@924175302 Example of SE
https://raw.githubusercontent.com/AlexeyAB/darknet/master/cfg/enet-coco.cfg
```
#squeeze-n-excitation
[avgpool]

# squeeze ratio r=16 (recommended r=16)
[convolutional]
filters=24
size=1
stride=1
activation=swish

# excitation
[convolutional]
filters=384
size=1
stride=1
activation=logistic

# multiply channels
[scale_channels]
from=-4
```
